src/pal/src/include/pal/virtual.h

```diff
@@ -105,6 +105,94 @@ BOOL VIRTUALOwnedRegion( IN UINT_PTR address );
+/// the reserved memory (or fails to allocate it during initialization) then VirtualAlloc/Reserve code
+/// will simply fall back to reserving memory using OS APIs.
+///
+/// Notes:
+///     - the memory allocated by this class is NOT committed by default. It is responsibility
+///       of the caller to commit the virtual memory before accessing it.
+///     - in addition, this class does not provide ability to free the reserved memory. The caller
+///       has full control of the memory it got from this allocator (i.e. the caller becomes
+///       the owner of the allocated memory), so it is caller's responsibility to free the memory
+///       if it is no longer needed.
+/// </summary>
+class ExecutableMemoryAllocator
+{
+public:
+    /// <summary>
+    /// This function initializes the allocator. It should be called early during process startup
+    /// (when process address space is pretty much empty) in order to have a chance to reserve
+    /// sufficient amount of memory that is close to the coreclr library.
+    /// </summary>
+    void Initialize();
+
+    /// <summary>
+    /// This function attempts to allocate the requested amount of memory from its reserved virtual
+    /// address space. The function will return NULL if the allocation request cannot
+    /// be satisfied by the memory that is currently available in the allocator.
+    /// </summary>
+    LPVOID AllocateMemory(int32_t allocationSize);
+
+private:
+    /// <summary>
+    /// This function is called during initialization. It opportunistically tries to reserve
+    /// a large chunk of virtual memory that can be later used to store JIT'ed code.
+    /// </summary>
+    void TryReserveInitialMemory();
+
+    /// <summary>
+    /// This function returns a random offset (in multiples of the virtual page size)
+    /// at which the allocator should start allocating memory from its reserved memory range.
+    /// </summary>
+    int32_t GenerateRandomStartOffset();
+
+private:
+    /// <summary>
+    /// There does not seem to be an easy way find the size of a library on Unix.
+    /// So this constant represents an approximation of the libcoreclr size (on debug build)
+    /// that can be used to calculate an approximate location of the memory that
+    /// is in 2GB range from the coreclr library. In addition, having precise size of libcoreclr
+    /// is not necessary for the calculations.
+    /// </summary>
+    const int32_t CoreClrLibrarySize = 100 * 1024 * 1024;
+
+    /// <summary>
+    /// This constant represent the max size of the virtual memory that this allocator
+    /// will try to reserve during initialization. We want all JIT-ed code and the
+    /// entire libcoreclr to be located in a 2GB range.
+    /// </summary>
+    const int32_t MaxExecutableMemorySize = 0x7FFF0000 - CoreClrLibrarySize;
+
+    /// <summary>Start address of the reserved virtual address space</summary>
+    LPVOID m_startAddress;
```

Comment:
The LPVOID seems to be the only windows style type used in this code. Could you please consider changing it to void*?